packages/server/src/simulation.ts

```diff
@@ -0,0 +1,97 @@
+import { Effect } from './effect';
+import express, { raw } from 'express';
+import { Behaviors, HttpApp, HttpHandler, HttpMethods, Methods, Protocols, Service, SimulationState, Simulator } from './interfaces';
+import { AddressInfo, createServer } from './http';
+
+
+export function simulation(definitions: Record<string, Simulator>): Effect<SimulationState> {
+  return slice => function*(scope) {
+    try {
+      let behaviors = Object.entries(definitions).reduce(({ services }, [name, simulator]) => {
+        return simulator(createBehaviors(services, name));
+      }, createBehaviors());
+
+      let servers = behaviors.services.map((service) => {
+        let app = express();
+        app.use(raw({ type: "*/*" }));
+        for (let handler of service.app.handlers) {
+          app[handler.method](handler.path, (request, response) => {
+            scope.spawn(function*() {
+              try {
+                yield handler.handler(request, response);
+              } catch(err) {
+                console.error(err);
+
+                response.status(500);
+                response.write('server error');
+              } finally {
+                response.end();
+              }
+            });
+          });
+        }
+
+        return {
+          name: service.name,
+          protocol: service.protocol,
+          server: createServer(app).run(scope)
+        };
+      });
+
+      let services: {name: string; url: string; }[] = [];
+      for (let { name, server, protocol } of servers) {
+        let address: AddressInfo = yield server.address();
+        services.push({ name, url: `${protocol}://localhost:${address.port}`});
+      }
+
+      slice.update(state => ({
+        ...state,
+        status: 'running',
+        services
+      }));
+
+      // all spun up, we can just wait.
+      yield;
+    } catch (error) {
+      slice.update((state) => ({
+        ...state,
+        status: "failed",
+        error
+      }));
+    }
+  };
+}
+
+function createBehaviors(services: Service[] = [], name = ''): Behaviors {
+  return {
+    services,
+    http(handler) {
+      let protocol: Protocols = 'http';
+      let app = handler(createHttpApp());
+      return createBehaviors(services.concat({ name, protocol, app }), name);
+    },
+    https(handler) {
+      let protocol: Protocols = 'https';
+      let app = handler(createHttpApp());
+      return createBehaviors(services.concat({ name, protocol, app }), name)
```

Comment:
does this cure the bug I mentioned previously where `services = []` every time `http` or `https` is called?

Might be worth having a test that adds more than 1 service and checks the length.

Reply:
I saw that you had fixed that, so I did that here too. The difference is that we call the constructor recursively so that the `services` array is rebound as a the next function parameter each time.

More testing sounds like a great idea.